### PR TITLE
KEYCLOAK-3303: Allow reuse of refresh tokens.

### PR DESCRIPTION
**Pull request by glavoie**

### CODE REVIEW

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java [hunk 1]
```diff
@@ -165,6 +165,9 @@ public static void importRealm(KeycloakSession session, RealmRepresentation rep,
         if (rep.getRevokeRefreshToken() != null) newRealm.setRevokeRefreshToken(rep.getRevokeRefreshToken());
         else newRealm.setRevokeRefreshToken(false);
 
+        if (rep.getRefreshTokenMaxReuse() != null) newRealm.setRefreshTokenMaxReuse(rep.getRefreshTokenMaxReuse());
+        else newRealm.setRefreshTokenMaxReuse(0);
```
Comment (glavoie) [inline server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java]: @mposolda, as discussed, default database value is not picked (at least by MySQL) for existing data as the column was set to NULLABLE to be in line with other setting columns. The default value is picked from here and it keeps the current behaviour of revoking a refresh token after first use.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticatedClientSessionAdapter.java [hunk 2]
```diff
@@ -156,6 +156,56 @@ public CrossDCMessageStatus getCrossDCMessageStatus(SessionEntityWrapper<UserSes
...
+
+            @Override
+            protected void runClientSessionUpdate(AuthenticatedClientSessionEntity entity) {
+                entity.setCurrentRefreshTokenUseCount(currentRefreshTokenUseCount);
+            }
+
+            @Override
+            public CrossDCMessageStatus getCrossDCMessageStatus(SessionEntityWrapper<UserSessionEntity> sessionWrapper) {
+                // We usually update lastSessionRefresh at the same time. That would handle it.
+                return CrossDCMessageStatus.NOT_NEEDED;
```
Comment (glavoie) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/AuthenticatedClientSessionAdapter.java]: Here, and in setCurrentRefreshToken I kept the same pattern as the timestamp as I understand that the update of the synchronization of the UserSessionEntity will synchronize the AuthenticatedClientSessionEntities by reference/cascade.

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java [hunk 3]
```diff
@@ -278,6 +270,36 @@ public RefreshResult refreshAccessToken(KeycloakSession session, UriInfo uriInfo
         return new RefreshResult(res, TokenUtil.TOKEN_TYPE_OFFLINE.equals(refreshToken.getType()));
     }
 
+    private void validateTokenReuse(KeycloakSession session, RealmModel realm, RefreshToken refreshToken,
+            TokenValidation validation) throws OAuthErrorException {
+        if (realm.isRevokeRefreshToken()) {
+            int clusterStartupTime = session.getProvider(ClusterProvider.class).getClusterStartupTime();
+
+            if (validation.clientSession.getCurrentRefreshToken() != null &&
```
Comment (glavoie) [inline services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java]: Ignoring this check if a token is being reused.

File: services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java [hunk 4]
```diff
@@ -278,6 +270,36 @@ public RefreshResult refreshAccessToken(KeycloakSession session, UriInfo uriInfo
...
+                    !refreshToken.getId().equals(validation.clientSession.getCurrentRefreshToken()) &&
+                    refreshToken.getIssuedAt() < validation.clientSession.getTimestamp() &&
+                    clusterStartupTime != validation.clientSession.getTimestamp()) {
+                throw new OAuthErrorException(OAuthErrorException.INVALID_GRANT, "Stale token");
+            }
+        }
+
+        if (realm.isRevokeRefreshToken()) {
```
Comment (glavoie) [inline services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java]: Manage the whole reuse logic here (updating token ID + counting).

**Comment by mposolda** [comment]
@glavoie Thanks again!

**Comment by glavoie** [comment]
Thank you @mposolda for the fast turnaround!

**Pull request merged**